Replace usage of NPM CLI with HTTP calls to the NPM Registry.

varungandhi-src opened the pull request:
I will be adding support for credentials in a subsequent PR.

Questions:
- [x] There is a self-enforced rate limit code; I'm not sure if that's needed? - Resolution: kept self-enforced rate limit
- [x] Naming: is `NPMWebClient` fine? I was thinking of naming it `NPMHTTPClient`, but that seemed like too many capital letters all at once. - Resolution: Renamed to `HTTPClient`.
- [x] Is there a way to expose a dependency for only testing (i.e. expose a type through a package's tests, making it available only to tests of other packages)? I read some SO answers and I don't think it's possible, and decided to put the `MockNPMClient` in a separate `npmtest` package, so that I could reuse it in tests in different packages. - Resolution: Kept separate `npmtest` package.
- [x] Should we be making actual NPM API calls in at least some of the tests here? Previously, we weren't making `npm` CLI calls; does that consideration change if we're using the API? - Resolution: Added record + replay tests for API calls.

mrnugget commented on internal/extsvc/npmpackages/npmtest/npmtest.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+type MockNPMClient struct {
+	T                       *testing.T
```
I can't come with a specific reason for why not to do this, but I think `*testing.T` behaves a bit like `context.Context` in that it is very context dependant and I'd personally not pass it around like that, but rather to functions/methods.

But that's not a hard "don't do this" (if it works, it works, I guess). When I look at the code though there's only one usage and that one _I think_ could also be an early-return of a custom error?

varungandhi-src replied:
Removed the `*testing.T` field.

mrnugget reviewed (approved):
Sweet! Left some comments on Go-y-ness of some code, but no huge blocker :)

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -7,50 +7,53 @@ import (
-	"path/filepath"
-	"regexp"
-	"strings"
+	"io"
```
Could you clarify what you mean here? There are two packages under `npmpackages`, one is `npm`, and the other `npmtest`. Are you suggesting moving both of them one level up: `internal/extsvc/npm` and `internal/extsvc/npmtest`?

mrnugget replied:
What I meant was to merge `npmpackages` and `npmpackages/npm` and rename it to `npm`. But that's just one option. 

Packages in Go are kinda hierarchical (see `net/http/httptest` for example) so `npmpackages/npm` strikes me as odd. And since there is, as far as I can tell, only one file in `npmpackages` (https://github.com/sourcegraph/sourcegraph/blob/main/internal/extsvc/npmpackages/repos.go) I don't see a reason for why that is the top-level package.

Instead we could do something like this, for example:

```
extsvc/npm
extsvc/npm/packages
extsvc/npm/npmtest
```

varungandhi-src replied:
Fixed by changing layout to

```
extsvc/npm
extsvc/npm/npmpackages
extsvc/npm/npmtest
```

I've kept the "npm" prefix in "npmpackages" to disambiguate at the point of use `npmpackages.Metadata` instead of `packages.Metadata`. The latter would be more natural for abstractions across different package ecosystems.

varungandhi-src commented on cmd/gitserver/server/vcs_syncer_npm_packages_test.go:
```diff
@@ -171,52 +179,10 @@ func TestNPMCloneCommand(t *testing.T) {
-	return npmPath.Name()
-}
-
 func (s NPMPackagesSyncer) runCloneCommand(t *testing.T, bareGitDirectory string, dependencies []string) {
```
It's a method because there (a) there is a corresponding method for `JVMPackagesSyncer` and (b) it is modifying the internal state of a private field `connection`, so it seems better suited to a method than a free function. I've tried to keep the code between the two largely the same (this PR actually creates more of a divergence than before, since we're doing API calls instead of shelling out), so that when we do a refactoring later (when a third package repo is added), it's easier to identify the similarities and differences.

Could you clarify what you mean by "not require overwriting non-test code from tests." ? I don't quite get what's being overwritten here.

varungandhi-src replied:
I've kept this method as-is for now; modifying private state feel more fitting as a method to me.

varungandhi-src commented on internal/repos/npm_packages.go:
```diff
@@ -86,8 +89,19 @@ func (s *NPMPackagesSource) ListRepos(ctx context.Context, results chan SourceRe
+			if !found {
+				versions, err = s.client.AvailablePackageVersions(ctx, npmDependency.Package)
+				if err != nil {
+					pkgVersions[pkgKey] = map[string]struct{}{}
```
Say I have `pkg@1.0`, `pkg@2.0`, and `pkg@3.0` in the database. In the first iteration of the loop, we'll ask NPM "what are the versions of `pkg` you've got?". Suppose we get an error. We don't need to repeat that API call again in the second iteration of the loop. Putting an empty map here means that the `versions, found := pkgVersions[pkgKey]` will return `found == true`, so we won't make API calls for 2.0 and 3.0 versions.

varungandhi-src commented on internal/extsvc/npmpackages/npmtest/npmtest.go:
```diff
@@ -0,0 +1,53 @@
+type MockNPMClient struct {
+	T                       *testing.T
+	TarballMap              map[string]string
+	DoesDependencyExistFunc *func(_ context.Context, dep reposource.NPMDependency) (exists bool, err error)
```
I hadn't realized `func` values were allowed to be `nil`. 😅

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -7,50 +7,53 @@ import (
-	npmConfigFetchTimeoutEnvVar = "NPM_CONFIG_FETCH_TIMEOUT"
-	npmConfigRegistryEnvVar     = "NPM_CONFIG_REGISTRY"
-)
+type NPMClient interface {
```
Fixed by renaming to `Client`.

varungandhi-src commented on cmd/gitserver/server/vcs_syncer_npm_packages.go:
```diff
@@ -310,32 +331,32 @@ func (s *NPMPackagesSyncer) commitTgz(ctx context.Context, dependency reposource
 	return action(tarReader)
 }
 
+type NamedReadSeeker struct {
```
Fixed by making it private.

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -7,50 +7,53 @@ import (
-	npmConfigRegistryEnvVar     = "NPM_CONFIG_REGISTRY"
-)
+type NPMClient interface {
+	// Find the available versions for an npm package.
```
I don't think that this convention is useful; it just makes docs read strangely. In other ecosystems, English documentation doesn't repeat the function's name. I don't see how "read better" applies to Go but not in other contexts.

* [C++](https://en.cppreference.com/w/cpp/algorithm/sort) - Sorts the elements in the range [first, last) in non-descending order.
* [Swift](https://developer.apple.com/documentation/swift/array/1688499-sort) - Sorts the collection in place.
* [Rust](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.sort) - Sorts the slice.
* [Go](https://pkg.go.dev/sort#Sort) - Sort sorts data.

I don't think the last one is better.

That said, I've added the names here because now that I've gotten the mini-rant out of my system, I am at peace. 😌

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -101,93 +85,156 @@ func Exists(ctx context.Context, config *schema.NPMPackagesConnection, dependenc
-		otlog.String("args", strings.Join(args, ", ")),
-	}})
-	defer endObservation(1, observation.Args{})
+type NPMWebClient struct {
```
Fixed by renaming to `HTTPClient`.

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -101,93 +85,156 @@ func Exists(ctx context.Context, config *schema.NPMPackagesConnection, dependenc
+		connection.Registry,
+		httpcli.ExternalDoer,
+		*rate.NewLimiter(rate.Limit(requestsPerHour/3600.0), 100),
+	}
```
I'm deliberately not using named fields to avoid someone accidentally adding a field to `NPMWebClient` (now `HTTPClient`) and forgetting to supply a sensible value in this function. If the linter did complain, the linter would be wrong and I would try to fix it. 😛

varungandhi-src commented on internal/extsvc/npmpackages/npm/npm.go:
```diff
@@ -101,93 +85,156 @@ func Exists(ctx context.Context, config *schema.NPMPackagesConnection, dependenc
+	return NPMWebClient{
+		connection.Registry,
+		httpcli.ExternalDoer,
+		*rate.NewLimiter(rate.Limit(requestsPerHour/3600.0), 100),
```
Fixed to use the same pattern as the BitBucker client.

varungandhi-src commented on internal/extsvc/npmpackages/npmtest/npmtest.go:
```diff
@@ -0,0 +1,53 @@
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/npmpackages/npm"
+)
+
+type MockNPMClient struct {
```
Dropped the NPM suffix.

Strum355 commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,242 @@
+	limiter     *rate.Limiter
+}
+
+func NewWebClient(connection schema.NPMPackagesConnection) HTTPClient {
```
Completely subjectively: I would personally split out the `connection.RateLimit` and `connection.Registry` into individual params here. Mostly a style preference

varungandhi-src replied:
Fixed by taking two parameters instead of one.

Strum355 commented on cmd/gitserver/server/vcs_syncer_npm_packages.go:
```diff
@@ -36,8 +36,28 @@ var (
+	if customClient != nil {
+		client = customClient
+	} else {
+		webClient := npm.NewWebClient(connection)
+		client = &webClient
```
`npm.NewWebClient` should return a pointer instead, given that all the methods are pointer methods and to avoid having this awkward address-taking dance required

varungandhi-src replied:
Fixed by returning a pointer in NewHTTPClient and NewNPMPackagesSyncer.

Strum355 commented on cmd/gitserver/main.go:
```diff
@@ -334,7 +334,8 @@ func getVCSSyncer(ctx context.Context, externalServiceStore database.ExternalSer
 			return nil, err
 		}
-		return &server.NPMPackagesSyncer{Config: &c, DBStore: codeintelDB}, nil
+		syncer := server.NewNPMPackagesSyncer(c, codeintelDB, nil)
+		return &syncer, nil
```
`server.NewNPMPackagesSyncer` should also return a pointer for the same reason as `npm.NewWebClient`. In general, returning pointers for structs would be the recommended "default" in cases like this, theres no avoiding the allocation because its going to escape the heap anyways by virtue of being passed around via an interface

Strum355 commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,242 @@
+
+func (client *HTTPClient) DoesDependencyExist(ctx context.Context, dep reposource.NPMDependency) (exists bool, err error) {
+	_, err = client.getDependencyInfo(ctx, dep)
+	var e illFormedJSONError
+	if errors.As(err, &e) {
+		return false, nil
+	}
```
Do we want to silently fail here or should we log `e`?

varungandhi-src replied:
Added a `log15.Warn` call for the error.

tsenart commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,242 @@
+
+func (client *HTTPClient) makeGetRequest(ctx context.Context, url string) (io.ReadSeeker, error) {
+	var bodyBuffer bytes.Reader
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, &bodyBuffer)
```
We're passing an empty request body here, and then returning it to the caller, which I think is a bug because we want to unmarshal the response body, not the request body. I don't think this method is very useful. We should just instantiate a request and pass it to do in `getDependencyInfo` and others.

varungandhi-src replied:
Thanks for catching that; I misunderstood the API originally. After fixing that bug, the method body is 15 lines long, and is called in 3 different places. Maybe I'm missing something; why do you think that this method isn't useful?

varungandhi-src replied:
> Would you also create makeHeadRequest, makePutRequest and makePostRequest?

I haven't created such methods because I don't need a more general API. If I needed a more general API, I would change this to take a request parameter.

> Seems duplicative if we already have a Do method and the only thing that changes between these is the request method

Currently, this method does 4 things:
1. Creates a request.
2. Calls `Do`.
3. Calls `io.Copy` (because I want an `io.ReadSeeker`, not just a `io.ReadCloser`).
4. Closes the response's `Body`.

It also has error checks for all those situations. If it were just a `Do` call, I would've been happy to inline it at the call-sites. However, the additional surrounding boilerplate means that if I duplicate this method at all the call-sites, I'll end up with 45 lines for what is currently ~15 lines. Moreover, the places where the body gets inlined will be harder to read because the code for making the request will be in the middle of more higher-level code that is concerned with application logic.

If you have a suggestion on how to simplify the method, I'm all ears. Similarly, if you have a suggestion on how to name it to be more idiomatic, I'm happy to change the method's name as well.

>  Overall, it doesn't feel very Go-ish.

For other similar comments, I've changed the code to be more idiomatic because there was not much downside. However, in this case, inlining the method body has a meaningful downside: code duplication. Given that, IMO "doesn't feel very Go-ish" is not a compelling enough reason to remove the method.

Strum355 commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return info, errors.Wrapf(err, "failed to read NPM's response: url=%s", url)
+	}
+	if json.Unmarshal(body, &info) != nil {
+		return info, illFormedJSONError{url: url}
+	}
```
Did I get the URL wrong or does `npmDependencyInfo` not match the JSON from the API? 
https://registry.npmjs.org/left-pad/1.3.0 
https://go.dev/play/p/4__QPOOvH8i

varungandhi-src replied:
No, you've got the right URL, I made a mistake; fixed now. I probably should've marked this PR as a draft since I hadn't done end-to-end testing again after making all the changes.

Strum355 commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,245 @@
+	_, err = client.getDependencyInfo(ctx, dep)
+	var e illFormedJSONError
+	if errors.As(err, &e) {
+		log15.Warn("received ill-formed JSON payload from NPM Registry API", "error", e)
```
This will log when a version doesn't exist as well (https://registry.npmjs.org/left-pad/1.3.1), we could probably the data handling here to avoid this

varungandhi-src replied:
The error payloads I'm seeing are:

```
# When package is not found
"Not found"
# When version of an existing package is not found
"version not found: 1.3.1"
```

In both cases, the result is not valid JSON. Is your suggestion that I could do string-matching against the payload and log a different error, or that I should reword the current message to cover all three cases (malformed JSON, package not found, version not found)?

mrnugget replied:
As far as I understand the API it uses HTTP status codes correctly, so why not check whether the status code is 404?

- https://registry.npmjs.org/thisdoesnotexist -> `404`
- https://registry.npmjs.org/foobar -> `200` (yes, it exists)
- https://registry.npmjs.org/foobar/1.1.0 -> `200`
- https://registry.npmjs.org/foobar/9.9.99 -> `404`

Strum355 replied:
That seems like a possibility, would need a lil reshuffling to expose the http code but sure

varungandhi-src replied:
Added a status code check in https://github.com/sourcegraph/sourcegraph/pull/29799/commits/4f8a755620b14708a1c53d487a9a6df459f79edf

I've kept an INFO level log if the package is not found, because it still seems like useful information to know. I'm not entirely sure what the difference is between DEBUG and INFO.

mrnugget replied:
> Added a status code check in 4f8a755

Just looked at that! Great! Three things:

1. I'd add the `Error() string` method to `npmErr` so it's a normal `error`
2. I wouldn't name it `npmErr` then but just `err`
3. Use a helper like [the `IsNotFound` here](https://github.com/sourcegraph/sourcegraph/blob/a118f07f3e07a773853287dae05e444c66dfbb98/internal/extsvc/github/common_test.go#L145-L147) to check what specific error it is (how api errors are handled in `github/v3.go` is quite exemplary for how we do it in other packages)

varungandhi-src replied:
^ Not doing this because even though it would be more idiomatic:

1. the `npmError` type does not cross package boundaries (the inner error is passed, that should already have the status code).
2. It requires down-casting, where that is not needed with the current code.
3. It requires more code overall but doesn't add any benefits IMO apart from being idiomatic.

Strum355 commented on internal/extsvc/npm/npm.go:
```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return nil, err
+	}
+	var bodyBuffer bytes.Buffer
+	if _, err := io.Copy(&bodyBuffer, resp.Body); err != nil {
+		return nil, err
+	}
```
I see that you read fully once here and again [here](https://github.com/sourcegraph/sourcegraph/pull/29799/files#diff-8d1981289649d679fd327ae672f3570dbe6390ed74a37e2df2c2c849e530df67R203). I feel we could probably avoid this by returning an `io.ReadCloser` (the `resp.Body`) directly. The fact its an `io.ReadCloser` should indicate that the calling function handles calling `.Close()`. What do you think?

varungandhi-src replied:
Good catch. I've removed the copy in the callers by returning a `[]byte` from `makeGetRequest` instead. I preferred that over returning an `io.ReadCloser` since the latter would require the read + close to happen at the 3 call-sites instead of in one place.

mrnugget commented:
Just noticed your question in the PR description:

> Should we be making actual NPM API calls in at least some of the tests here? Previously, we weren't making npm CLI calls; does that consideration change if we're using the API?

I think so, yes. We have other client tests that record HTTP calls when the tests are ran with a flag (`-update`, for example). Those, I feel, are often good to weed out some easy-to-make mistakes (unmarshaling JSON the wrong way, for example). See: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httptestutil/recorder.go?L47:6&subtree=true#tab=references and https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httptestutil/recorder.go?L94:6&subtree=true#tab=references

varungandhi-src commented:
Temporarily marked the PR as Do not merge as I just discovered an issue with getting the right package versions from the API. Investigating that.

varungandhi-src commented:
I've added one test in https://github.com/sourcegraph/sourcegraph/pull/29799/commits/7cb7ef87f7bbfbfeffae0f4a0e8f7d74e5dc0d61 using the NPM registry API and the recording+replaying as suggested. @mrnugget does that look directionally correct? If so, I'm happy to add similar test cases for other functions as well.

mrnugget commented:
Yup! Looks good! Left a comment on the bit about the `ExternalClientFactory`. And also: you might want to use the existing factory for this?https://github.com/sourcegraph/sourcegraph/blob/7cb7ef87f7bbfbfeffae0f4a0e8f7d74e5dc0d61/internal/httptestutil/recorder.go#L94

varungandhi-src commented:
> Or was there a specific reason for using a new factory (and switchign away from ExternalDoer?

> Yup! Looks good! Left a comment on the bit about the ExternalClientFactory. And also: you might want to use the existing factory for this?

Thanks, I hadn't realized I could swap out the doer with a doer that handles record+replay. SInce I couldn't figure out how to use the API related to `NewRecorderOpt` and I didn't notice the recording factory API, I took a look at the [go-vcr](https://github.com/dnaeon/go-vcr) docs and figured I needed a `http.Client`. So I removed the Doer and replaced it with a Client, and implemented the same logic.

This simplifies things quite a bit.

I will submit a separate PR later with some doc comments which I think could be useful.
